### PR TITLE
Add UI page for social login

### PR DESCRIPTION
**Pull request by HungNgien**
Add UI page to allow toggle on/off sign-in options
image

image

## Summary by Sourcery

Add a new UI page for managing social login settings, allowing administrators to toggle sign-in options for various providers. Integrate 'django-allauth' to support social authentication with providers like Google, GitHub, and MediaWiki. Enhance the system with OAuth2 application management capabilities, including creation and deletion. Update tests to accommodate the new social login URLs.

New Features:
- Introduce a new UI page for managing social login settings, allowing administrators to enable or disable sign-in options for different providers.

Enhancements:
- Add support for OAuth2 applications, including creation and deletion functionalities, to facilitate social login integration.

Build:
- Integrate the 'django-allauth' library to support social authentication, including providers like Google, GitHub, and MediaWiki.

Documentation:
- Add user-facing documentation for the new social login settings page, explaining how to enable or disable login providers.

Tests:
- Update test infrastructure to ignore social URLs from django-allauth, as they do not support namespace.

## Summary by Sourcery

Add a UI page for managing social login settings, allowing administrators to enable or disable different providers.

New Features:
- Added a new user interface for enabling and disabling social login providers.

Tests:
- Updated tests to exclude social login URLs.

### CODE REVIEW

**Comment by mariobehling** [comment]
Always add screenshots of changes when making a PR, please!

**Comment by HungNguyen-tma** [comment]
> Always add screenshots of changes when making a PR, please!

Sorry, I was just preparing the draft pr, I'm resolving some conflicts.

File: src/pretix/plugins/socialauth/templates/socialauth/social_auth_settings.html [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                <legend>{% trans "Social login providers" %}</legend>
+                <div class="table-responsive">
+                    <table class="table">
+                        {% for provider, enable in login_providers.items %}
+                        <tr class="{% if enable %}success{% else %}default{% endif %}">
+                            <td>
+                                <strong>{% trans provider %}</strong>
+                                <p>{% trans "Login with " %}{{ provider|capfirst }}</p>
+                            </td>
+                            <td class="text-right flip" width="20%">
```
Comment (odkhang) [inline src/pretix/plugins/socialauth/templates/socialauth/social_auth_settings.html:36]: avoid using css inline if possible

File: src/pretix/plugins/socialauth/views.py [hunk 2]
```diff
@@ -28,3 +46,37 @@ def oauth_return(request):
...
+        context['login_providers'] = self.gs.settings.get('login_providers', as_type=dict)
+        return context
+
+    def post(self, request, *args, **kwargs):
+        login_providers = self.gs.settings.get('login_providers', as_type=dict)
+        for provider in self.LOGIN_PROVIDERS.keys():
+            value = request.POST.get(f'{provider}_login', '').lower()
+            if value not in self.VALID_STATES:
+                continue
+            elif value == "enable":
```
Comment (odkhang) [inline src/pretix/plugins/socialauth/views.py]: try not to use raw text "enable" here

**Comment by mariobehling** [comment]
Please make button "Login with Email" for the default option and when the user clicks the button expand the view and show standard login as below.
Screenshot from 2024-12-20 05-54-35

**Comment by mariobehling** [comment]
So, at first it should look something like that

397286361-d9395597-2fd4-4025-875d-924c2f78ce3a

**Comment by HungNguyen-tma** [comment]
image
image
@mariobehling, does it match your idea? Click the button "Login with email" will expand or hide the login form

**Comment by mariobehling** [comment]
Thanks, yes, looks good.

Please correct spelling mistakes on settings pages - Mediawiki, Github, Google.

Can you confirm Internationalization works here? Please translate to Vietnamese as a sample language.

File: src/pretix/plugins/socialauth/urls.py [hunk 3]
```diff
@@ -3,5 +3,7 @@
 from . import views
 
 urlpatterns = [
-    url(r'^oauth_return$', views.oauth_return, name='social.oauth.return')
+    url(r'^oauth_login/(?P<provider>[a-zA-Z]+)/$', views.oauth_login, name='social.oauth.login'),
```
Comment (hongquan) [inline src/pretix/plugins/socialauth/urls.py]: There is no reason to use old method for new code.
Follow this new style instead: https://docs.djangoproject.com/en/5.1/topics/http/urls/

**Review by hongquan — CHANGES_REQUESTED** [review]

File: src/pretix/plugins/socialauth/views.py [hunk 4]
```diff
@@ -28,3 +47,38 @@ def oauth_return(request):
         )
         logger.error('Error while authorizing: user has no email address.')
         return redirect('control:auth.login')
+
+
+class LoginState(Enum):
```
Comment (hongquan) [inline src/pretix/plugins/socialauth/views.py]: Please subclass from [`StrEnum`](https://docs.python.org/3.11/library/enum.html#enum.StrEnum). By that, you can compare directly with `str`, like `v === LoginState.ENABLED`.

File: src/pretix/plugins/socialauth/views.py [hunk 5]
```diff
@@ -28,3 +47,38 @@ def oauth_return(request):
...
+    ENABLE = "enable"
```
Comment (hongquan) [inline src/pretix/plugins/socialauth/views.py]: Renamed to past participle, like "ENABLED", "DISABLED".

**Comment by HungNguyen-tma** [comment]
> Thanks, yes, looks good.
> 
> Please correct spelling mistakes on settings pages - Mediawiki, Github, Google.
> 
> Can you confirm Internationalization works here? Please translate to Vietnamese as a sample language.

image
I asked @odkhang to help me deploy this on test server, and it works. Must be the error in my local setup.
Vietnamese is not supported in this src, I tested it with German instead.

File: src/pretix/plugins/socialauth/views.py [hunk 6]
```diff
@@ -28,3 +47,37 @@ def oauth_return(request):
...
+        if self.gs.settings.get('login_providers', as_type=dict) is None:
+            self.gs.settings.set('login_providers', self.LOGIN_PROVIDERS)
+        context['login_providers'] = self.gs.settings.get('login_providers', as_type=dict)
+        return context
+
+    def post(self, request, *args, **kwargs):
+        login_providers = self.gs.settings.get('login_providers', as_type=dict)
+        for provider in self.LOGIN_PROVIDERS.keys():
+            value = request.POST.get(f'{provider}_login', '').lower()
+            if value not in [s.value for s in LoginState]:
```
Comment (hongquan) [inline src/pretix/plugins/socialauth/views.py:76]: As stated in previous comment, if `LoginState` is a subclass  of `StrEnum`, its member can compare directly with `str`.
So this check can be as simple as `if value not in LoginState`.

image

**Review by hongquan — CHANGES_REQUESTED** [review]

File: src/pretix/plugins/socialauth/views.py (hunk 6 above)
Comment (HungNguyen-tma) [reply src/pretix/plugins/socialauth/views.py:76]: @hongquan, it looks like this syntax is only working for Python 3.12 or later.
I checked it on my python3.11 and an online compiler, both raise TypeError: unsupported operand type(s) for 'in': 'str' and 'EnumType'.

**Pull request merged**